core/src/test/java/io/grpc/PickFirstLoadBalancer2Test.java

```diff
@@ -0,0 +1,235 @@
+  @Captor
+  private ArgumentCaptor<Picker> pickerCaptor;
+  @Captor
+  private ArgumentCaptor<Attributes> attrsCaptor;
+  @Mock
+  private Helper mockHelper;
+  @Mock
+  private Subchannel mockSubchannel;
+
+  @Before
+  public void setUp() {
+    MockitoAnnotations.initMocks(this);
+    for (int i = 0; i < 3; i++) {
+      SocketAddress addr = new FakeSocketAddress("server" + i);
+      servers.add(ResolvedServerInfoGroup.builder().add(new ResolvedServerInfo(addr)).build());
+      socketAddresses.add(addr);
+    }
+
+    when(mockSubchannel.getAddresses()).thenReturn(new EquivalentAddressGroup(socketAddresses));
+    when(mockSubchannel.getAttributes()).thenReturn(Attributes.newBuilder()
+        .set(LAST_STATE, new AtomicReference<ConnectivityState>(IDLE))
+        .build());
+    when(mockHelper.createSubchannel(any(EquivalentAddressGroup.class), any(Attributes.class)))
+        .thenReturn(mockSubchannel);
+
+    loadBalancer = (PickFirstBalancer) PickFirstBalancerFactory2.getInstance().newLoadBalancer(
+        mockHelper);
+  }
+
+  @Test
+  public void pickAfterResolved() throws Exception {
+    loadBalancer.handleResolvedAddresses(servers, affinity);
+
+    verify(mockHelper).createSubchannel(eagCaptor.capture(), attrsCaptor.capture());
+    verify(mockHelper).updatePicker(pickerCaptor.capture());
+
+    assertEquals(new EquivalentAddressGroup(socketAddresses), eagCaptor.getValue());
+    assertEquals(pickerCaptor.getValue().pickSubchannel(affinity, new Metadata()),
+        pickerCaptor.getValue().pickSubchannel(affinity, new Metadata()));
+
+    verifyNoMoreInteractions(mockHelper);
+  }
+
+  @Test
+  public void pickAfterResolvedAndUnchanged() throws Exception {
+    loadBalancer.handleResolvedAddresses(servers, affinity);
+    loadBalancer.handleResolvedAddresses(servers, affinity);
+
+    verify(mockHelper, times(1)).createSubchannel(any(EquivalentAddressGroup.class),
+        any(Attributes.class));
+    verify(mockHelper, times(1)).updatePicker(isA(Picker.class));
+
+    verifyNoMoreInteractions(mockHelper);
+  }
+
+  @Test
+  public void pickAfterResolvedAndChanged() throws Exception {
+    SocketAddress socketAddr = new FakeSocketAddress("newserver");
+    List<SocketAddress> newSocketAddresses = Lists.newArrayList(socketAddr);
+    List<ResolvedServerInfoGroup> newServers = Lists.newArrayList(
+        ResolvedServerInfoGroup.builder().add(new ResolvedServerInfo(socketAddr)).build());
+
+    final Subchannel oldSubchannel = mock(Subchannel.class);
+    final EquivalentAddressGroup oldEag = new EquivalentAddressGroup(socketAddresses);
+    when(oldSubchannel.getAddresses()).thenReturn(oldEag);
+
+    final Subchannel newSubchannel = mock(Subchannel.class);
+    final EquivalentAddressGroup newEag = new EquivalentAddressGroup(newSocketAddresses);
+    when(newSubchannel.getAddresses()).thenReturn(newEag);
+
+    when(mockHelper.createSubchannel(any(EquivalentAddressGroup.class), any(Attributes.class)))
```

Comment:
Right, I think it was doing something complicated initially. Thanks for catching.